plugins/eventmonitor/eventmonitor.cpp

```diff
@@ -0,0 +1,187 @@
+        return "QChildEvent";
+    case QEvent::Timer:
+        return "QTimerEvent";
+    case QEvent::MetaCall:
+        return "QMetaCallEvent";  // about to change in 5.14? see https://code.qt.io/cgit/qt/qtbase.git/commit/?h=dev&id=999c26dd83ad37fcd7a2b2fc62c0281f38c8e6e0
+    default:
+        return "";
+    }
+}
+
+
+static bool eventCallback(void **data)
+{
+    QEvent *event = reinterpret_cast<QEvent*>(data[1]);
+    QObject *receiver = reinterpret_cast<QObject*>(data[0]);
+
+    if (!event || !receiver) {
+        qWarning() << "Event or receiver is invalid";
+        return false;
+    }
+
+    if (Probe::instance()->filterObject(receiver)) {
+        return false;
+    }
+
+    EventData eventData;
+    eventData.time = QTime::currentTime();
+    eventData.type = event->type();
+    eventData.receiver = receiver;
+    eventData.attributes << QPair<const char*, QVariant>{"receiver", QVariant::fromValue(receiver)};
+
+    QString className = eventTypeToClassName(event->type());
+    if (!className.isEmpty()) {
+        MetaObject *metaObj = MetaObjectRepository::instance()->metaObject(className);
+        if (metaObj) {
+            for (int i=0; i<metaObj->propertyCount(); ++i) {
+                MetaProperty* prop = metaObj->propertyAt(i);
+                if (strcmp(prop->name(), "type") == 0) continue;
+                eventData.attributes << QPair<const char*, QVariant>{prop->name(), prop->value(event)};
+            }
+        }
+    }
+
+    if (s_model) {
+        // add directly from foreground thread, delay from background thread
+        QMetaObject::invokeMethod(s_model, "addEvent", Qt::AutoConnection,
+                                  Q_ARG(GammaRay::EventData, eventData));
+    }
+    return false;
+}
+
+
+EventMonitor::EventMonitor(Probe *probe, QObject *parent)
+    : QObject(parent)
+    , m_eventModel(new EventModel(this))
+    , m_eventPropertyModel(new AggregatedPropertyModel(this))
+{
+    Q_ASSERT(s_model == nullptr);
+    s_model = m_eventModel;
+
+    QInternal::registerCallback(QInternal::EventNotifyCallback, eventCallback);
+
+    auto proxy = new ServerProxyModel<QSortFilterProxyModel>(this);
+    proxy->setSourceModel(m_eventModel);
+    probe->registerModel(QStringLiteral("com.kdab.GammaRay.EventModel"), proxy);
+
+    probe->registerModel(QStringLiteral("com.kdab.GammaRay.EventPropertyModel"), m_eventPropertyModel);
+
+    QItemSelectionModel *selectionModel = ObjectBroker::selectionModel(proxy);
+    connect(selectionModel, &QItemSelectionModel::selectionChanged,
+            this, &EventMonitor::eventSelected);
+}
+
+void EventMonitor::eventSelected(const QItemSelection &selection)
+{
+    if (selection.isEmpty())
+        return;
+    const QModelIndex index = selection.first().topLeft();
+    QVariant eventAttributes = QVariant(index.data(EventModelRole::AttributesRole).value<QVariantMap>());
+
+    m_eventPropertyModel->setObject(eventAttributes);
+}
+
+EventMonitor::~EventMonitor() = default;
```

Comment:
This probably should probably reset s_model to nullptr and unregister the callback. This doesn't matter on application exit, but you can detach the probe from a running application and continue without GammaRay (as well as re-attach later to such a process), for that cleaning up especially also "static" state is important.